### PR TITLE
[BEAM-7427] Fix JmsCheckpointMark Avro Encoding

### PR DESCRIPTION
**Pull request by zouabimourad**
`JmsCheckpointMark` can't be serialized with AvroCoder beause of the inner class `State`
Avro schema generation of `JmsCheckpointMark` fails
the externalisation of the inner class fixes the issue.

R: @iemejia 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by iemejia** [comment]
Thanks for bringing a fix for this one. Can you run `./gradlew spotlessApply` to fix the checkstyle issue and re push again. Thanks.

File: sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java [hunk 1]
```diff
@@ -333,6 +335,13 @@ public void testCheckpointMark() throws Exception {
     assertEquals(0, count(QUEUE));
   }
 
+  @Test
+  public void testJmsCheckpointMarkAvroEncoding() {
+    AvroCoder<JmsCheckpointMark> avroCoder = AvroCoder.of(JmsCheckpointMark.class);
+
+    assertNotNull(avroCoder);
```
Comment (iemejia) [inline sdks/java/io/jms/src/test/java/org/apache/beam/sdk/io/jms/JmsIOTest.java]: Can you please add a test here that validates that we can encode and decode an instance of  `JmsCheckpointMark` correctly. For example by using: `CoderProperties.coderDecodeEncodeEqual(avroCoder, checkpointMark);` or `CoderProperties.structuralValueDecodeEncodeEqual(avroCoder, checkpointMark);`

**Comment by zouabimourad** [comment]
> IMHO, as original author of this IO, I would refactor the checkpoint mark to use `JmsRecord` instead of `Message` and ack on this. It would straight forward without wrapping.
> @zouabimourad let me know if you want me to do the change or you wanna try.

If you have better fix .. please commit it

**Comment by jbonofre** [comment]
@zouabimourad ok, let me propose a fix then.

**Comment by iemejia** [comment]
@zouabimourad just for some extra context my comment on testing encoding/decoding comes from the fact that `CheckpointMark`s should be able to be converted to bytes. However the current implementation is trying to do that for a list of not yet processed `javax.jms.Message`s which is probably not a good idea because we cannot guarantee that those will be correctly serialized with or without Avro. So probably a better approach (and if I understand correctly is what @jbonofre proposes) is to contain a list of `JmsRecord` in `JmsCheckpointMark` because we know those are `Serializable`.

**Comment by zouabimourad** [comment]
> @zouabimourad just for some extra context my comment on testing encoding/decoding comes from the fact that `CheckpointMark`s should be able to be converted from/to bytes. However the current implementation is trying to do that for a list of not yet processed `javax.jms.Message`s which is probably not a good idea because we cannot guarantee that those will be correctly serialized with or without Avro. So probably a better approach (and if I understand correctly is what @jbonofre proposes) is to contain a list of `JmsRecord` in `JmsCheckpointMark` because we know those are `Serializable`.

I agree with you `javax.jms.Message` could not be deserialized since it's an interface .. i got errors after using `CoderProperties.coderDecodeEncodeEqual(avroCoder, checkpointMark);` in test.

@jbonofre has a better fix.

**Comment by jbonofre** [comment]
I'm working on it.

**Comment by sdotz** [comment]
Any progress on this? Currently stuck on this bug. Thanks!

**Comment by aaltay** [comment]
Hi all, could we move forward with the PR as is or do we expect additional changes here? If additional changes are required could we do them in a follow up PR? (I would like to understand if we can unblock @sdotz.)

**Comment by sdotz** [comment]
> I would like to understand if we can unblock @sdotz.

I took the code from the PR and am using it. So far so good. Consider me unblocked, but it would still be nice to get some closure and official solution one way or another.

**Comment by kolban-google** [comment]
@sdotz Howdy sir ... I have a customer who seems to be suffering this puzzle.  Am curious on your post that you took the PR code and are using it.  Assuming that someone like me doesn't have the skills to re-build from a PR is there a recipe that can be followed to achieve what you did?  Is there a replacement JAR or other "piece" that you may be able to offer up "as-is" for those who want to try this fix?

**Comment by kolban-google** [comment]
I built the patch using this Pull Request and initially thought that it had solved the puzzle.  However, very quickly I found that I was failing with [BEAM-4409](https://issues.apache.org/jira/browse/BEAM-4409).  As I now read this Pull Request in detail, I seem to see that y'all are aware of this too.  The last I seem to see is from mr @jbonofre back in 2019-09-04 where there is a comment that it is being worked upon.  Are there any updates on this?  Any approximate timeframe for a patch to test?

**Comment by jbonofre** [comment]
Hi all, sorry I forgot to move forward on this one.

I'm resuming my work on it. Sorry for the delay.

**Comment by jbonofre** [comment]
I'm evaluating the following options:
0. Add a test with AvroCoder on checkpoint mark
1. use JmsRecord in the checkpoint instead of Message to ack. The issue is how to "load" Message from JmsRecord.
2. use Session to ack instead of message. The Session stores the unack messages, so the checkpoint mark could delegate to the session.

I'm evaluating the different options. I keep you posted soon.

**Comment by jbonofre** [comment]
I think I will use the same approach as I did in MQTT with transient. I'm testing it.

**Comment by jbonofre** [comment]
Superseded by #10644 10644

**Comment by iemejia** [comment]
For info #10644 was merged today. The fix will be part of Beam 2.20.0 since the vote for 2.19.0 has already started.

**Comment by tannaga** [comment]
@iemejia Do you happen to know when Beam 2.20.0 will be released? Is there a snapshot version with this fix I can try locally to see if it fixes the issue I am running into?

**Comment by iemejia** [comment]
You can test it right now by using the 2.20.0-SNAPSHOT version available in Apache. The release branch will be cut next wednesday. Usually it takes us around 2 weeks after that to get the release voted and out.

**Pull request closed without merging**